src/Polly.Strategies/Internals/PredicatesHandler.cs

```diff
@@ -0,0 +1,38 @@
+﻿namespace Polly.Internals;
+
+internal abstract class PredicatesHandler<TContext>
+{
+    public abstract ValueTask<bool> HandleAsync<T>(Outcome<T> outcome, TContext context, ResilienceContext resilienceContext);
+
+    public static PredicatesHandler<TContext> Create(Predicates<TContext> input) => new Many(input.Callbacks);
+
+    private class Many : PredicatesHandler<TContext>
+    {
+        private readonly Dictionary<Type, List<object>> _callbacks;
+
+        public Many(Dictionary<Type, List<object>> callbacks)
+        {
+            _callbacks = callbacks;
+        }
+
+        public override async ValueTask<bool> HandleAsync<T>(Outcome<T> outcome, TContext context, ResilienceContext resilienceContext)
+        {
+            if (!_callbacks.TryGetValue(typeof(T), out var predicates))
+            {
+                return false;
+            }
+
+            foreach (var predicate in predicates)
+            {
+                var outcomePredicate = (PredicatesCallback<T, TContext>)predicate;
+
+                if (await outcomePredicate(outcome, context).ConfigureAwait(resilienceContext.ContinueOnCapturedContext))
```

Comment:
Should we defensively cast with an expression (so we could ignore/skip), or should it never not cast so blowing up is fine?

Reply:
It should always case fine, because we have the setup fully in our control. The caller can only put supported types there (and is always stored as `PredicatesCallback<T, TContext>`).